File: membrain-py/membrain/rerankers/openai.py
"""OpenAI LLM-based reranker."""

from __future__ import annotations

import time

from ..types import RerankResults, SearchResults
from .base import (
    DEFAULT_TIMEOUT,
    LLM_SYSTEM_PROMPT,
    BaseReranker,
    build_llm_user_prompt,
    build_rerank_results,
    http_post,
    parse_llm_scores,
)

_OPENAI_CHAT_URL = "https://api.openai.com/v1/chat/completions"
_OPENAI_DEFAULT_MODEL = "gpt-4o-mini"


class OpenAIReranker(BaseReranker):
    """LLM-based reranker using the OpenAI Chat Completions API.

    Supports any chat model: gpt-4o-mini, gpt-4o, etc.
    """

    def __init__(
        self,
        api_key: str,
        model: str = _OPENAI_DEFAULT_MODEL,
        top_k: int = 5,
        endpoint: str | None = None,
        timeout: int = DEFAULT_TIMEOUT,
    ) -> None:
        self.api_key = api_key
        self.model = model
        self.top_k = top_k
        self.endpoint = endpoint or _OPENAI_CHAT_URL
        self.timeout = timeout

    def rerank(
        self,
        query: str,
        results: SearchResults,
        top_k: int | None = None,
    ) -> RerankResults:
        effective_top_k = top_k if top_k is not None else self.top_k
        memories = results.memories
        if not memories:
            return RerankResults(model=self.model, provider="openai")

        documents = [memory.content for memory in memories]
        user_prompt = build_llm_user_prompt(query, documents)

        start = time.monotonic()
        response = http_post(
            url=self.endpoint,
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json",
            },
            body={
                "model": self.model,
                "messages": [
                    {"role": "system", "content": LLM_SYSTEM_PROMPT},
                    {"role": "user", "content": user_prompt},
                ],
                "temperature": 0.0,
            },
            timeout=self.timeout,
        )
        duration_ms = int((time.monotonic() - start) * 1000)

        response_text = (
            response.get("choices", [{}])[0]
            .get("message", {})
            .get("content", "[]")
        )
        scored_indices = parse_llm_scores(
            response_text, len(documents), effective_top_k
        )

        return build_rerank_results(
            memories, scored_indices, self.model, "openai", duration_ms
        )

File: membrain-node/src/rerankers/openai.ts
/**
 * OpenAI LLM-based reranker.
 */

import type { RerankerConfig, SearchResults, RerankResults } from "../types";

import {
  BaseReranker,
  DEFAULT_TIMEOUT,
  LLM_SYSTEM_PROMPT,
  httpPost,
  buildRerankResults,
  buildLlmUserPrompt,
  parseLlmScores,
} from "./base";

const OPENAI_CHAT_URL = "https://api.openai.com/v1/chat/completions";
const OPENAI_DEFAULT_MODEL = "gpt-4o-mini";

export class OpenAIReranker extends BaseReranker {
  private readonly apiKey: string;
  private readonly model: string;
  private readonly topK: number;
  private readonly endpoint: string;
  private readonly timeout: number;

  /**
   * LLM-based reranker using the OpenAI Chat Completions API.
   *
   * Supports any chat model: gpt-4o-mini, gpt-4o, etc.
   */
  constructor(config: RerankerConfig) {
    super();
    this.apiKey = config.apiKey;
    this.model = config.model ?? OPENAI_DEFAULT_MODEL;
    this.topK = config.topK ?? 5;
    this.endpoint = config.endpoint ?? OPENAI_CHAT_URL;
    this.timeout = config.timeout ?? DEFAULT_TIMEOUT;
  }

  async rerank(
    query: string,
    results: SearchResults,
    topK?: number,
  ): Promise<RerankResults> {
    const effectiveTopK = topK ?? this.topK;
    const memories = results.memories;
    if (memories.length === 0) {
      return { memories: [], model: this.model, provider: "openai", duration_ms: 0 };
    }

    const documents = memories.map((memory) => memory.content);
    const userPrompt = buildLlmUserPrompt(query, documents);

    const start = performance.now();
    const response = await httpPost(
      this.endpoint,
      {
        Authorization: `Bearer ${this.apiKey}`,
        "Content-Type": "application/json",
      },
      {
        model: this.model,
        messages: [
          { role: "system", content: LLM_SYSTEM_PROMPT },
          { role: "user", content: userPrompt },
        ],
        temperature: 0.0,
      },
      this.timeout,
    );
    const durationMs = Math.round(performance.now() - start);

    const responseText =
      response.choices?.[0]?.message?.content ?? "[]";
    const scoredIndices = parseLlmScores(
      responseText,
      documents.length,
      effectiveTopK,
    );

    return buildRerankResults(
      memories,
      scoredIndices,
      this.model,
      "openai",
      durationMs,
    );
  }
}
